tensorflow/python/autograph/operators/py_builtins.py

```diff
@@ -278,6 +280,10 @@ def raise_zero_rank_error():
                                raise_zero_rank_error)
 
 
+def _tf_dataset_len(s):
+  return s.reduce(constant_op.constant(0, dtype=dtypes.int32), lambda x, _: x + 1)
```

Comment:
Please use the cardinality function instead, it can be much more efficient: https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/data/experimental/ops/cardinality.py#L33
It can also protect against infinite loops (repeated datasets will cycle forever).

If the function returns INFINITE, then we should raise an error.

If it returns UNKNOWN, then we have two options:
 * raise an error outright (safer)
 * fall back to counting with reduce, but in that case we'd need to limit the number of iterations in the eventuality that the dataset might be actually infinite